lib/ohai/util/file_helper.rb

```diff
@@ -16,23 +16,39 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-# Copied from chef/lib/chef/util/selinux.rb
-
 module Ohai
   module Util
     module FileHelper
       def which(cmd)
-        paths = ENV["PATH"].split(File::PATH_SEPARATOR) + [ "/bin", "/usr/bin", "/sbin", "/usr/sbin" ]
-        paths.each do |path|
-          filename = File.join(path, cmd)
-          if File.executable?(filename)
+        path_cmd = (collect_os == :windows) ? "echo %PATH%" : "echo $PATH"
+        path = connection.run_command(path_cmd).stdout
+        paths = path.split(File::PATH_SEPARATOR) + %w{/bin /usr/bin /sbin /usr/sbin}
+
+        paths.each do |pathdir|
+          filename = File.join(pathdir, cmd)
+          backend_file_stat = connection.file(filename).stat
+
+          # Train does not return stat information for Windows yet
+          next if collect_os != :windows && backend_file_stat.empty?
+
+          # Blanket accept for Windows or octal mode check for others
+          if collect_os == :windows || (backend_file_stat[:mode] & 0111) > 0
             logger.trace("Plugin #{name}: found #{cmd} at #{filename}")
             return filename
           end
         end
+
         logger.trace("Plugin #{name}: did not find #{cmd}")
         false
       end
+
+      def file_exist?(path)
+        connection.file(path).exist?
+      end
```

Comment:
So instead of this, ohai should be using chef-util to get these helper functions:

https://github.com/chef/chef/blob/master/chef-utils/lib/chef-utils/dsl/train_helpers.rb

There's also a train-compatible which helper:

https://github.com/chef/chef/blob/master/chef-utils/lib/chef-utils/dsl/which.rb

Those should get injected here, with some "dependency injection" to wire up the location of the
train transport.  What I came up with is slightly janky but I documented it here:

https://github.com/chef/chef/tree/master/chef-utils#documentation-for-software-developers

The next big bit is that the shell_out helper inside of chef-client itself needs to get moved into chef-utils (and some of the tight coupling to Chef::Config and Chef::Log (and any other internal constructs, the timeout stuff in particular my be problematic) needs to be made optional / dep-injected), then the implementation of shell_out in ohai needs to go away and it just needs to get it from chef-utils:

https://github.com/chef/chef/blob/master/lib/chef/mixin/shell_out.rb

So for example you probably want to have a `def __config` similarly to `def __transport_connection` in here:

https://github.com/chef/chef/blob/master/chef-utils/lib/chef-utils/internal.rb

I don't know what that is by default, you may have to raise NotImplementedError and then in the chef-client classes override that to be Chef::Config (otherwise you take a circular dep on the Chef class).  You can try to use the same trick I did with protecting the Chef.run_context and having the Chef stuff be wired up automatically, but honestly I'm starting to hate that now that I'm looking at it with fresh eyes.  That injection wiring probably needs to be moved into the chef gem and everything there should throw some notimplerror.

So anyway then in ohai you'd include the helpers into a module in ohai, then override `def __config` and def __transport_connection`.

Reply:
I'm going to go do some work now on the chef-side to try to extract the shell_out mixin properly

Reply:
This was one of the concerns I wrote down reading this over this morning. We want to make sure we sane paths added when we shell out or Ohai is going to stop returning valid data on EL platforms. We either need to hack that back into this helper or we can use chef-utils to get that via which.

Reply:
https://github.com/chef/chef/pull/10137